fixing definition of CXX11_ABI for gcc-like

memsharded opened the pull request:
Changelog: Bugfix: Fix the definition of D_GLIBCXX_USE_CXX11_ABI in gcc-like compilers for ``CMakeToolchain`` and ``AutotoolsToolchain``. Define it only to ``D_GLIBCXX_USE_CXX11_ABI=0`` for new compilers, assuming that the default is alread 1.
Docs: Omit

Address https://github.com/conan-io/conan-center-index/pull/8389

Lets see if something breaks, need to add tests.

memsharded committed:
fixing definition of CXX11_ABI for gcc-like

memsharded committed:
fix test

memsharded commented on conan/tools/cmake/toolchain.py:
```diff
@@ -187,11 +187,9 @@ def context(self):
-            elif libcxx == "libstdc++":
+
+        if compiler in ['clang', 'apple-clang', 'gcc']:
+            if libcxx == "libstdc++":
```
Lets be aware that from https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

```
Using the default configuration options for GCC the default value of the macro is 1 which causes the new ABI to be active, so to use the old ABI you must explicitly define the macro to 0 before including any library headers. (Be aware that some GNU/Linux distributions configure GCC 5 differently so that the default value of the macro is 0 and users must define it to 1 to enable the new ABI.)
```
So if we go this way, users from those distros will need to customize the ``CMakeToolchain``, something like this:

```python
def generate(self):
    tc = CMakeToolchain(self)
    tc.blocks["libcxx"].values["glibcxx"] = "1" if self.settings.libcxx == "libstdc++11" else "0"
    tc.generate()
```

There were requests to eliminate the default ``GLIBCXX_USE_CXX11_ABI=1``, as it was raising some warnings (undefined in some compilers, but not errors), so it seems there is no perfect solution for 100% of users.

lasote reviewed (approved):
I would add a test with the `generate` you suggest in case someone uses a custom distro. Also, we have to document this default.

memsharded committed:
added conf and tests

memsharded commented:
> I would add a test with the generate you suggest in case someone uses a custom distro. Also, we have to document this default.

I have added tests for both ``CMakeToolchain`` and ``AutotoolsToolchain``, probably needs to be added to Meson, but I see no trace of libcxx there, so that would need more time.

I have also added a new conf ``tools.gnu:define_libcxx11_abi``, as it seems it would be very inconvenient for users of ConanCenter recipes, if they don't want to fork and modify those recipes, for example. What do you think?

lasote reviewed (approved):
👍  Could be merged
